File: firebase/firestore/lib/src/sections/reader.dart
import "package:firestore/helpers.dart";

/// A collection of functions to read course data.
/// 
/// No function in this class actually performs logic on said data, just returns
/// it. This helps keep the program modular, by separating the data sources from
/// the data indexing.
class SectionReader {
	/// Maps course IDs to their respective names.
	static Future<Map<String, String>> get courseNames async => {
		await for (final Map<String, String> row in csvReader(DataFiles.courses))
			if (row ["SCHOOL_ID"] == "Upper")
				row ["ID"]: row ["FULL_NAME"]
	};

	/// Maps section IDs to their respective faculty IDs. 
	// static Future<Map<String, String>> getSectionFacultyIds() async {
	// 	final Map<String, String> result = {};
	// 	await for(final Map<String, String> row in csvReader(DataFiles.section)) {
	// 		final String teacher = row ["FACULTY_ID"];
	// 		if (row ["SCHOOL_ID"] != "Upper" || teacher.isEmpty) {
	// 			continue;
	// 		}
	// 		result [row ["SECTION_ID"]] = teacher;
	// 	}
	// 	return result;
	// } 

	static Future<Map<String, String>> getSectionFacultyIds() async => {
		await for (final Map<String, String> row in csvReader(DataFiles.section))
			if (row ["SCHOOL_ID"] == "Upper" && row ["FACULTY_ID"].isNotEmpty)
				row ["SECTION_ID"]: row ["FACULTY_ID"],
	};
}

File: firebase/firestore-py/lib/sections/reader.py
import csv
from ..utils import dir, logger

'''
A collection of functions to read course data.

No function in this class actually performs logic on data, just returns it.
his helps keep the program modular, by separating the data sources from
the data indexing.
'''
# fixes ids of form "90070.0" to look like "90070"
# but allows ids of CADV
def fix_id(id):
  try:
    return str(int(float(id)))
  except ValueError:
    return id
    
def get_course_names():
  with open(dir.courses) as file:
    return {
      fix_id(row["Course ID"]): row["Course Name"]
      for row in csv.DictReader(file) 
      if row["School ID"] == "Upper"}
  
def get_section_faculty_ids():
  with open(dir.section) as file: 
    return {
      row["SECTION_ID"]: row["FACULTY_ID"]
      for row in csv.DictReader(file)
      if row["SCHOOL_ID"] == "Upper" and row["FACULTY_ID"]}
  

def get_zoom_links():
  Links = {}
  try:
    with open(dir.zoom_links) as file:
      for row in csv.DictReader(file):
        if row["LINK"]:
          Links[row["EMAIL"]] = row["LINK"]

  except FileNotFoundError:
    logger.warning("zoom_links.csv doesn't exist. Cannot grab data. Using an empty dictionary instead")

  try:
    with open(dir.special_zoom_links) as file:
      for row in csv.DictReader(file):
        Links[row["ID"]] = row["LINK"]

  except FileNotFoundError:
    logger.warning("No special zoom links")
  
  return Links

